File: typescript/src/data-structures/deque.ts
/**
 * Double-ended queue implementation using a doubly linked list.
 * Time: O(1) for all operations
 * Space: O(n)
 */

interface DequeNode<T> {
  value: T;
  prev: DequeNode<T> | null;
  next: DequeNode<T> | null;
}

export class Deque<T> {
  private head: DequeNode<T> | null = null;
  private tail: DequeNode<T> | null = null;
  private _size: number = 0;

  pushFront(value: T): void {
    const node: DequeNode<T> = { value, prev: null, next: this.head };
    if (this.head) {
      this.head.prev = node;
    } else {
      this.tail = node;
    }
    this.head = node;
    this._size++;
  }

  pushBack(value: T): void {
    const node: DequeNode<T> = { value, prev: this.tail, next: null };
    if (this.tail) {
      this.tail.next = node;
    } else {
      this.head = node;
    }
    this.tail = node;
    this._size++;
  }

  popFront(): T | undefined {
    if (!this.head) {
      return undefined;
    }
    const value = this.head.value;
    this.head = this.head.next;
    if (this.head) {
      this.head.prev = null;
    } else {
      this.tail = null;
    }
    this._size--;
    return value;
  }

  popBack(): T | undefined {
    if (!this.tail) {
      return undefined;
    }
    const value = this.tail.value;
    this.tail = this.tail.prev;
    if (this.tail) {
      this.tail.next = null;
    } else {
      this.head = null;
    }
    this._size--;
    return value;
  }

  peekFront(): T | undefined {
    return this.head?.value;
  }

  peekBack(): T | undefined {
    return this.tail?.value;
  }

  isEmpty(): boolean {
    return this._size === 0;
  }

  size(): number {
    return this._size;
  }

  clear(): void {
    this.head = null;
    this.tail = null;
    this._size = 0;
  }
}

File: swift/Sources/DSA/DataStructures/Deque.swift
/// Double-ended queue supporting operations at both ends.
/// Time: O(1) for all operations
/// Space: O(n)
public struct Deque<T> {
    private class Node {
        var value: T
        var prev: Node?
        var next: Node?

        init(_ value: T, prev: Node? = nil, next: Node? = nil) {
            self.value = value
            self.prev = prev
            self.next = next
        }
    }

    private var head: Node?
    private var tail: Node?
    private var size = 0

    public init() {}

    public mutating func pushFront(_ value: T) {
        let node = Node(value, next: head)
        if let head = head {
            head.prev = node
        } else {
            tail = node
        }
        head = node
        size += 1
    }

    public mutating func pushBack(_ value: T) {
        let node = Node(value, prev: tail)
        if let tail = tail {
            tail.next = node
        } else {
            head = node
        }
        tail = node
        size += 1
    }

    public mutating func popFront() -> T? {
        guard let node = head else { return nil }
        head = node.next
        if let head = head {
            head.prev = nil
        } else {
            tail = nil
        }
        size -= 1
        return node.value
    }

    public mutating func popBack() -> T? {
        guard let node = tail else { return nil }
        tail = node.prev
        if let tail = tail {
            tail.next = nil
        } else {
            head = nil
        }
        size -= 1
        return node.value
    }

    public func peekFront() -> T? {
        head?.value
    }

    public func peekBack() -> T? {
        tail?.value
    }

    public var isEmpty: Bool {
        size == 0
    }

    public var count: Int {
        size
    }

    public mutating func clear() {
        head = nil
        tail = nil
        size = 0
    }
}
